src/libcore/hash/mod.rs

```diff
@@ -56,186 +53,382 @@
 //! let person1 = Person { id: 5, name: "Janet".to_string(), phone: 555_666_7777 };
 //! let person2 = Person { id: 5, name: "Bob".to_string(), phone: 555_666_7777 };
 //!
-//! assert!(hash::hash(&person1) == hash::hash(&person2));
+//! assert_eq!(hash::<_, SipHasher>(&person1), hash::<_, SipHasher>(&person2));
 //! ```
 
-#![allow(unused_must_use)]
+#![unstable = "module was recently redesigned"]
 
-use prelude::*;
+use default::Default;
 
-use borrow::{Cow, ToOwned};
-use intrinsics::TypeId;
-use mem;
-use num::Int;
+pub use self::sip::SipHasher;
 
-/// Reexport the `sip::hash` function as our default hasher.
-pub use self::sip::hash as hash;
+mod sip;
 
-pub mod sip;
+/// A hashable type.
+///
+/// The `H` type parameter is an abstract hash state that is used by the `Hash`
+/// to compute the hash.
+#[cfg(stage0)]
+pub trait Hash<H> {
+    /// Feeds this value into the state given, updating the hasher as necessary.
+    fn hash(&self, state: &mut H);
+}
 
-/// A hashable type. The `S` type parameter is an abstract hash state that is
-/// used by the `Hash` to compute the hash. It defaults to
-/// `std::hash::sip::SipState`.
-pub trait Hash<S = sip::SipState> {
-    /// Computes the hash of a value.
-    fn hash(&self, state: &mut S);
+/// A hashable type.
+///
+/// The `H` type parameter is an abstract hash state that is used by the `Hash`
+/// to compute the hash.
+#[cfg(not(stage0))]
+pub trait Hash<H: Hasher> {
+    /// Feeds this value into the state given, updating the hasher as necessary.
+    fn hash(&self, state: &mut H);
 }
 
-/// A trait that computes a hash for a value. The main users of this trait are
-/// containers like `HashMap`, which need a generic way hash multiple types.
-pub trait Hasher<S> {
-    /// Compute the hash of a value.
-    fn hash<T: ?Sized + Hash<S>>(&self, value: &T) -> u64;
+/// A trait which represents the ability to hash an arbitrary stream of bytes.
+pub trait Hasher {
+    /// Result type of one run of hashing generated by this hasher.
+    type Output;
+
+    /// Resets this hasher back to its initial state (as if it were just
+    /// created).
+    fn reset(&mut self);
+
+    /// Completes a round of hashing, producing the output hash generated.
+    fn finish(&self) -> Self::Output;
 }
 
+/// A common bound on the `Hasher` parameter to `Hash` implementations in order
+/// to generically hash an aggregate.
+#[experimental = "this trait will likely be replaced by io::Writer"]
 #[allow(missing_docs)]
 pub trait Writer {
     fn write(&mut self, bytes: &[u8]);
 }
 
+/// Hash a value with the default SipHasher algorithm (two initial keys of 0).
+///
+/// The specified value will be hashed with this hasher and then the resulting
+/// hash will be returned.
+pub fn hash<T: Hash<H>, H: Hasher + Default>(value: &T) -> H::Output {
```

Comment:
Yeah, although there may perhaps be some better methods of providing a function like this:

* We could provide a default method on the `Hash` trait with a `where Hasher: Default` clause is specified and put this function there. That way you'd say something like `foo.hash_value::<Algorithm>()`
* I *think* that with Niko's default type parameter RFC we could swap the order of `T` and `H`, and you would only have to specify `H`, right? That is, `foo::<H>` is equivalent to `foo::<H, _>`?

This function isn't super useful now that the default `SipHasher` bound was removed, so it may be better to just remove it.